rfd/0-job-worker.md

```diff
@@ -0,0 +1,240 @@
+
+##### Job Output Streaming
+
+The worker library streams job output through an `io.ReadCloser`. The server will indicate the end of the stream, if the job has completed, with the `io.EOF` error. The client may, at any time, close the stream to indicate that it is disconnecting. The client must always close the stream when it no longer needs it to prevent memory leaks.
+
+Internally, the library will maintain a buffer containing the job output. The job itself will be configured to use this buffer for `stdout` and `stderr` and the `Write()` method will be goroutine safe. When a client calls `JobOutput()` the library will create a new `io.Reader` that independently, and with goroutine safety, reads through to the end of the output buffer. If the job has already completed at this time, `io.EOF` will be returned. If not, subsequent calls to `Read()` will block until there is either more output to return or the job ends in which case `io.EOF` is returned. When the job calls `Write()` it will signal connected readers that new data has been written to the output buffer. The readers will in turn be able to unblock and return their `Read()` calls to connected clients at that time.
+
+##### Stopping Jobs
+
+StopJob is an asynchronous request that takes an optional timeout. If the timeout is greater than 0, it will cause the library to first issue a `SIGINT` and wait up to `timeout` for the job to complete. If the job doesn't complete after `timeout`, or timeout was `0`, then the job is terminated with `SIGKILL`. The library returns a channel that will be closed when the job completes.
+
+##### Identifiers
+
+JobID is an opaque identifier that is internally mapped to job in the library. Since pid is not guaranteed to be unique, nothing internal to the library will be keyed off of the pid. The library must still track the pid though in order to signal the process to stop, for example. The `go.jetify.com/typeid` library is used to generate job ids.
+
+UserID can be any unique value as far as the library is concerned. See the section on Authorization below for more information.
+
+#### gRPC API
+
+The gRPC API follows the worker library fairly closely. The only significant difference is that it does not require the user id in the messages because user id is the client certificate's subject, not explicitly provided. Additionally, `StopJob()` in the library returns a channel that closes when the job completes. The gRPC API does not provide a facility for notifying the client when the job completes.
+
+### Security Considerations
+
+#### Authentication
+
+mTLS is used for authentication. The server will be configured with a ca certificate that it expects all client certificates to be signed by. All connections presenting a certificate, signed by the configured ca certificate, are authenticated, all other connections are not.
```

Comment:
* Are we using the same CA for both server and clients?
* Could you call out the certificate algorithm / key length?

Reply:
* technically, it doesn't have to be, but the certs i've generated to test this do use the same ca for both. i'll be sure to commit the certs, keys and generation script (presuming i can figure out how to get past the secrets scanning ci) when it's time.
* i generated my examples keys with `secp256r1` and signed the certs with sha256

Reply:
@codingllama i just realized you may be asking for the cert algo / key type to be included in the design. was this your intent, or were you just curious in the comments here?

Reply:
Yes, it's best to call out the algo / key length in the design.

(Plus Tim's question.)

Reply:
> Are there any security concerns from using the same CA for clients and servers?

For production services, you'd generally want to use a server cert signed by a CA that the clients already trust (either a public CA or because you manage the client and installed a trusted root CA).

As for the client cert, while public CAs do offer client cert signing services, it's usually for S/MIME email. Besides, with the way the server is configured for the challenge, we want to be restrictive in issuing client certs, so we control who's authenticated. For that reason we'd probably want to use our own CA to issue client certs.

To your question, are there security concerns with using the same CA? If we were using a public CA then anyone could get a signed certificate that the server would authenticate, so yes.

Separately, it would be a good idea to use separate CAs, even if they're intermediate CAs both signed by the same root CA (one for server certs and another for client certs). This way, if one is compromised (assuming the root cert is managed much more securely than the intermediates, e.g. offline hardware device) then only some of the certs have to be reissued.

Reply:
updated the design doc with pki security recommendations